File: src/main/kotlin/fr/jhelp/kotlinLigthTools/promise/FutureTools.kt
package fr.jhelp.kotlinLigthTools.promise

fun <T> unwrap(future: Future<Future<T>>): Future<T>
{
    val promise = Promise<T>()

    future.then { futureCombined ->
        when (futureCombined.status())
        {
            FutureStatus.SUCCEED -> promise.result(futureCombined.getResult()!!.getResult()!!)
            FutureStatus.FAILED -> promise.error(future.getError())
            FutureStatus.CANCELED -> promise.getFuture().cancel(future.getCancelReason()!!)
            else                  ->
            {
            }
        }
    }

    return promise.getFuture()
}

fun <T> futureValue(value:T) : Future<T>
{
    val promise = Promise<T>()
    promise.result(value)
    return promise.getFuture()
}

fun <T> futureFailed(error:String) : Future<T>
{
    val promise = Promise<T>()
    promise.error(error)
    return promise.getFuture()
}

File: src/main/swift/Sources/swift/fr/jhelp/kotlinLigthTools/promise/FutureTools.swift

func unwrap <T>(_ future: Future<Future<T>>) -> Future<T>
{
    let promise = Promise<T>()

    future.then { futureCombined in 
        switch futureCombined.status()
        {
            case FutureStatus.SUCCEED  :  promise.result(futureCombined.getResult()!.getResult()!)
            case FutureStatus.FAILED  :  promise.error(future.getError())
            case FutureStatus.CANCELED  :  promise.getFuture().cancel(future.getCancelReason()!)
            default                   : 
            do {
            }
        }
    }

    return promise.getFuture()
}

func futureValue <T>(_ value:T) -> Future<T>
{
    let promise = Promise<T>()
    promise.result(value)
    return promise.getFuture()
}

func futureFailed <T>(_ error:String) -> Future<T>
{
    let promise = Promise<T>()
    promise.error(error)
    return promise.getFuture()
}
